kernel/src/platform/snp.rs

```diff
@@ -140,60 +133,52 @@ impl SvsmPlatform for SnpPlatform {
             return Err(SvsmError::NotSupported);
         }
 
-        self.use_alternate_injection = alt_inj_requested;
+        APIC_EMULATION_REG_COUNT.store(1, Ordering::Relaxed);
         Ok(())
     }
 
-    fn use_alternate_injection(&self) -> bool {
-        self.use_alternate_injection
-    }
-
-    fn lock_unlock_apic_emulation(&self, lock: bool) -> Result<(), SvsmError> {
-        // The lock state can only be changed if APIC emulation has not already
-        // been disabled on any CPU.
-        let new_state = if lock {
-            APIC_EMULATION_LOCKED
-        } else {
-            APIC_EMULATION_ENABLED
-        };
-        let mut current = APIC_EMULATION_STATE.load(Ordering::Relaxed);
+    fn change_apic_registration_state(&self, incr: bool) -> Result<bool, SvsmError> {
+        let mut current = APIC_EMULATION_REG_COUNT.load(Ordering::Relaxed);
         loop {
-            if current == APIC_EMULATION_DISABLED {
-                return Err(SvsmError::Apic);
-            }
-            match APIC_EMULATION_STATE.compare_exchange_weak(
+            let new = if incr {
+                // Incrementing is only possible if the registration count
+                // has not already dropped to zero, and only if the
+                // registration count will not wrap around.
+                if current == 0 {
+                    return Err(SvsmError::Apic);
+                }
+                match current.checked_add(1) {
+                    None => {
+                        return Err(SvsmError::Apic);
+                    }
+                    Some(v) => v,
+                }
```

Comment:
Just as a suggestion:

```suggestion
                current.checked_add(1).ok_or(SvsmError::Apic)?
```